File: lib/repositories/note_repository.dart
import '../db/database_helper.dart';
import '../models/note.dart';
import '../services/notification_service.dart';

enum SortMode { modified, created, alphabetical, color }

class NoteRepository {
  final _dbHelper = DatabaseHelper.instance;

  Future<List<Note>> getNotes({
    required bool archived,
    required bool deleted,
    String query = '',
    SortMode sortMode = SortMode.modified,
  }) async {
    final db = await _dbHelper.database;
    final maps = await db.query(
      'notes',
      where: 'isArchived = ? AND isDeleted = ?',
      whereArgs: [archived ? 1 : 0, deleted ? 1 : 0],
    );
    var notes = maps.map((m) => Note.fromMap(m)).toList();

    if (query.trim().isNotEmpty) {
      final q = query.trim().toLowerCase();
      notes = notes.where((n) => n.searchableText.contains(q)).toList();
    }

    switch (sortMode) {
      case SortMode.modified:
        notes.sort((a, b) => b.modifiedAt.compareTo(a.modifiedAt));
        break;
      case SortMode.created:
        notes.sort((a, b) => b.createdAt.compareTo(a.createdAt));
        break;
      case SortMode.alphabetical:
        notes.sort((a, b) => a.title.toLowerCase().compareTo(b.title.toLowerCase()));
        break;
      case SortMode.color:
        notes.sort((a, b) => a.colorIndex.compareTo(b.colorIndex));
        break;
    }
    return notes;
  }

  Future<List<Note>> getNotesWithReminders() async {
    final db = await _dbHelper.database;
    final maps = await db.query('notes', where: 'isDeleted = 0 AND reminderAt IS NOT NULL');
    return maps.map((m) => Note.fromMap(m)).toList();
  }

  Future<void> insert(Note note) async {
    final db = await _dbHelper.database;
    await db.insert('notes', note.toMap());
  }

  Future<void> update(Note note) async {
    final db = await _dbHelper.database;
    note.modifiedAt = DateTime.now();
    await db.update('notes', note.toMap(), where: 'id = ?', whereArgs: [note.id]);
  }

  Future<void> setArchived(String id, bool value) async {
    final db = await _dbHelper.database;
    await db.update('notes', {'isArchived': value ? 1 : 0, 'modifiedAt': DateTime.now().millisecondsSinceEpoch},
        where: 'id = ?', whereArgs: [id]);
  }

  Future<void> setDeleted(String id, bool value) async {
    final db = await _dbHelper.database;
    await db.update('notes', {'isDeleted': value ? 1 : 0, 'modifiedAt': DateTime.now().millisecondsSinceEpoch},
        where: 'id = ?', whereArgs: [id]);
  }

  Future<void> setLocked(String id, bool value) async {
    final db = await _dbHelper.database;
    await db.update('notes', {'isLocked': value ? 1 : 0}, where: 'id = ?', whereArgs: [id]);
  }

  Future<void> permanentDelete(String id) async {
    final db = await _dbHelper.database;
    await db.delete('notes', where: 'id = ?', whereArgs: [id]);
  }

  Future<void> emptyTrash() async {
    final db = await _dbHelper.database;
    final trashed = await db.query('notes', columns: ['id'], where: 'isDeleted = 1');
    for (final row in trashed) {
      await NotificationService().cancelReminder(row['id'] as String);
    }
    await db.delete('notes', where: 'isDeleted = 1');
  }
}

File: app/src/main/java/com/jotter/notes/data/NoteRepository.kt
package com.jotter.notes.data

import kotlinx.coroutines.flow.Flow

enum class SortMode { MODIFIED, CREATED, ALPHABETICAL, COLOR }

class NoteRepository(private val dao: NoteDao) {

    fun observeActive(): Flow<List<Note>> = dao.observe(archived = false, deleted = false)
    fun observeArchived(): Flow<List<Note>> = dao.observe(archived = true, deleted = false)
    fun observeTrashed(): Flow<List<Note>> = dao.observe(archived = false, deleted = true)
    fun observeWithReminders(): Flow<List<Note>> = dao.observeWithReminders()

    fun sortAndFilter(notes: List<Note>, query: String, sortMode: SortMode): List<Note> {
        var result = notes
        if (query.isNotBlank()) {
            val q = query.trim().lowercase()
            result = result.filter { it.searchableText().contains(q) }
        }
        result = when (sortMode) {
            SortMode.MODIFIED -> result.sortedByDescending { it.modifiedAt }
            SortMode.CREATED -> result.sortedByDescending { it.createdAt }
            SortMode.ALPHABETICAL -> result.sortedBy { it.title.lowercase() }
            SortMode.COLOR -> result.sortedBy { it.colorIndex }
        }
        return result
    }

    suspend fun save(note: Note) {
        note.modifiedAt = System.currentTimeMillis()
        dao.upsert(note)
    }

    suspend fun archive(id: String) = dao.setArchived(id, true)
    suspend fun unarchive(id: String) = dao.setArchived(id, false)
    suspend fun trash(id: String) = dao.setDeleted(id, true)
    suspend fun restore(id: String) = dao.setDeleted(id, false)
    suspend fun permanentDelete(id: String) = dao.permanentDelete(id)
    suspend fun setLocked(id: String, value: Boolean) = dao.setLocked(id, value)
    suspend fun getById(id: String): Note? = dao.getById(id)
}
